### PR TITLE
Data Manager RFC

### PR DESCRIPTION
**Pull request by Pessimistress**
Trying to generalize the need that arose from pydeck's binary support.

### CODE REVIEW

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+```js
+converter.convert(json, buffers);
+```
+
+Where `buffers` is a list/map of ArrayBuffers.
+
+In `json`, binary data can be referenced using the following syntax:
+
+```
+@@binary[<id>]/<type>/<offset>/<length>
```
Comment (twojtasz) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: This feels in spirit similar to the buffer, bufferViews, accessors of gltf binary format
https://kcoley.github.io/glTF/specification/2.0/figures/gltfOverview-2.0.0a.png

**Comment by ajduberstein** [comment]
This looks fine to me–happy to implement it in @deck.gl/json and include it in pydeck.

**Comment by twojtasz** [comment]
I'm expanding on my thoughts behind the gltf similarities with the goal being to broaden or create a record of the consideration rather than suggest we move to gltf.  It has many graphics related features, but there is a big difference adopting that in it's entirety which is beyond what we need for deck.gl.

**tl;dr:**
The current proposal is very specific to providing some mechanism for binary data and has a very tight scope.  If we open that scope a bit i would want to consider:

 1. binary blob that contains multiple actual binary arrays (single request for all data)
 2. ability to encode binary data as a URI, making it part of the JSON rather than a side-car
 3. flexibility & need for interleaved data in the binary buffer

These issues are addressed in **gltf** by expanding and separating out how the binary metadata.  Rather than simple and concise format of `@@binary[1]/uint8`, it breaks this down into a few arrays of structured metadata related to the buffers, views into the buffer, and type information etc.

Nothing would prohibit moving forward with this proposal as is and expanding to this more flexible structure definition when necessary.

**--- Verbose section**

Having said that the goals for this RFC seem to be:

1. expand the json work to support binary encoded data
2. secondarily consider environments not limited to the browser (pydeck being python for example)

This proposal assume the json & binary data are provided.  It may be useful include a workflow where **just** the json is provided but the **binary** data could be linked to by a URL.  This provides some nice developer experience, where data could simply be linked and let the system handle fetching and hydrating.

This line of thinking then would take influence from the **gltf** where they break down the "binary" description into actual structured elements, rather than the condensed structured string proposed here.

Specifically it would look something like this:

```config = {
   layers: [
       {
            "@@type": "PointCloudLayer",
            "data": {
            "length": 10000,
            "attributes": {
              "getPosition": "@@accessor[0]",
              "getColor": "@@accessor[1]"
            }
        }
   ],
   buffers: [
        { byteLength: 500, uri: "binary.bin" }
   },
   bufferViews: [
      {
           buffer: 0,
           byteOffset: 0,
           byteLength: 400
      },
      {
           buffer: 0,
           byteOffset: 400,
           byteLength: 100
      }
   ],
   accessor: [
       { 
            bufferView: 0
            type: "VEC2",
            componentType: 5126
       }
    ]
}
```

This it not a complete counter-proposal, but hopefully enough that you can see where / why gltf did what it did and if there is anything we could leverage to balance the flexibility vs overhead of their approach.

https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_005_BuffersBufferViewsAccessors.md

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+```
+
+- `id`: key in `buffers`
+- `type`: one of `int8`, `uint8`, `int16`, `uint16`, `int32`, `uint32`, `float32`, `float64`
+- `offset` (optional): byte offset into the buffer, default `0`
+- `length` (optional): byte length of the data, default to end of buffer
```
Comment (tsherif) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: This type of parameter is usually called `size` ([e.g.](https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glBufferSubData.xml)).

**Comment by tsherif** [comment]
This approach seems sensible. This architecture of creating a "database" of buffers and referring to them by an identifier is common in 3D engines, primarily for the purpose of re-using buffers.

As an alternative to the glTF model @twojtasz suggested, we could frame it along the lines of WebGPU, where [buffer instances](https://github.com/tsherif/webgpu-examples/blob/gh-pages/triangle.html#L72-L92), [their usage](https://github.com/tsherif/webgpu-examples/blob/gh-pages/triangle.html#L138-L157), and [the mapping between the two](https://github.com/tsherif/webgpu-examples/blob/gh-pages/triangle.html#L180-L181) are handled separately.

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+## Proposal
+
+Expand the `JSONConverter.convert` API to accept a second argument:
+
+```js
+converter.convert(json, buffers);
+```
+
+Where `buffers` is a list/map of ArrayBuffers.
```
Comment (ibgreen) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: Sending individual buffers is absolutely the reasonable starting point, though from an assumed typical user persona's perspective this could be considered an advanced use case. 

For binary data, it could be desirable to work towards supporting binary formats that contain all buffers and some or all of metadata required to access them) in a single file, i.e. Arrow (and possibly glTF), so that they could just be sent as a unit and then just supplied directly as `Layer.data`.

The canonical example being a pydeck user just needs to use `pyarrow` to convert his `pandas` data frame to a binary Arrow blob and then pass that blob as data to a deck.gl layer. Ideally the user has to provide as little mapping as possible on the deck.gl side. (The in-progress Arrow support RFCs have the beginnings of the required data mappings).

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- `offset` (optional): byte offset into the buffer, default `0`
+- `length` (optional): byte length of the data, default to end of buffer
+
+For example:
+
+```json
+{
+  "@@type": "PointCloudLayer",
+  "data": {
+    "length": 10000,
```
Comment (ibgreen) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: Ideas for the longer term: To make it a little easier on the user it could perhaps be nice to be able to let the user assemble the buffer "soup" into a data source that could be referenced in the layers. Rather than have all these defs inline in the layer. Though that would require starting to think about the data sources API.
```
"dataSources": {
   "bufferTable": {
     buffer1: {props},
```

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 1 above)
Comment (ibgreen) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: Nit: Is there a strong reason to have a separate text syntax here for the type, instead of reusing the accessor props from the JS API? Thinking about the high-level "One API" principle...

The `GL` constants should be part of the configuration.

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+- Updates: It is not possible to update a layer without re-sending the entire snapshot. This is important for animation and/or live data visualization.
+
+A hybrid transport format of both JSON and binary will help address these issues.
+
+## Proposal
+
+Expand the `JSONConverter.convert` API to accept a second argument:
+
+```js
+converter.convert(json, buffers);
```
Comment (ibgreen) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: There is not much detail on how buffers are transported, which is somewhat reasonable given that we may have widely varying implementations on different platforms. 

But we probably need a buffer or data source manager (perhaps on the @deck.gl/json level) and it probably needs to have a defined API that the implementations can work against.

They requirement is of course that buffers can be sent separately from JSON and that JSON can be updated multiple times, referencing the same buffers.

File: dev-docs/RFCs/proposals/transport-format-rfc.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+```js
+let json;
+let dataBuffer = <ring buffer>;
+
+function onMessage({data}) {
+  if (data.type === 'json') {
+    json = data.data;
+  } else if (data.type === 'binary') {
+    dataBuffer.push(data);
```
Comment (ibgreen) [inline dev-docs/RFCs/proposals/transport-format-rfc.md]: A potential question is also if named buffers can be replaced (e.g. allowing animation/transitions by sending new buffers).

It looks like it is possible here though it seems like old buffers would never be released?

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 4 above)
Comment (twojtasz) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: This is along the lines of the gltf or webGPU structure right.

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 5 above)
Comment (twojtasz) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: ```They requirement is of course that buffers can be sent separately from JSON and that JSON can be updated multiple times, referencing the same buffers.```

This part seems to be an important requirement not really covered here.  The idea is that there is a set of state (JSON and Buffers) and you can update both or one, and the "correct" behavior should happen.

Does this also mean supporting partial updates?  For example, I want to only update 1 buffer out of the 10 that where sent.

Comment (ibgreen) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: Yes look at the message sequence diagram in the linked slides. Ideally there should be the ability to send independent messages with data "blobs", and independent messages with JSON styles. 

I am envisioning e.g. a pydeck user working interactively with pyarrow, creating binary blobs, sending them to deck, then sending new JSON layers  referencing those new blobs to update the animation via e.g. transitions.

And yes, for that kind of use case it would be cool if "named" data blobs could be replaced (or as you say "partially updated", i.e. not just added.

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 6 above)
Comment (Pessimistress) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: They can definitely be released/replaced. The idea is that the **application** should control buffer management, not deck. In this code sample, `let dataBuffer = <ring buffer>;` the application is using a ring buffer implementation to store buffers that it receives.

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 1 above)
Comment (Pessimistress) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: I agree. I will update the proposal.

File: dev-docs/RFCs/proposals/transport-format-rfc.md (hunk 3 above)
Comment (Pessimistress) [reply dev-docs/RFCs/proposals/transport-format-rfc.md]: Per my understanding of the current frameworks architecture, we don't want deck.gl to parse the Arrow format. Deck should understand the table category from loaders.gl. It's up to the application what format they want to use, and they can pick the corresponding loader, which outputs to the same, normalized format.

If that is the case, then there is no built-in support for "a single binary file." Deck will always receive a JSON descriptor, which contains accessors and arraybuffers. It is not deck's concern how a single binary is parsed to that JSON descriptor.

**Comment by kylebarron** [comment]
> Per my understanding of the current frameworks architecture, we don't want deck.gl to parse the Arrow format

Is this still true? Just jotting down a few possible benefits to looking into Arrow for the binary transport format:

**Apply accessors on JS side**

The provided example is:
```js
{
  "@@type": "PointCloudLayer",
  "data": {
    "length": 10000,
    "attributes": {
      "getPosition": "@@binary[0]/float64",
      "getColor": "@@binary[1]/uint8>"
  }
}
```

This implies to me that layer accessors are always resolved in the binding instead of in Deck.gl. E.g. here `getColor` accesses RGB values directly, so if the user wants to apply a _function_ to generate colors, that function _must_ be applied in the binding.

That means that any time a user wants to update the color or any other attribute based on a function, the binding must apply the new function, and then send a _new binary buffer_. If a new buffer must be sent on every style update, that seems to minimize part of the benefit of separating the JSON and the data in the first place. 

I could envision passing a binary Arrow table from the binding to JS, and then possibly serializing a function definition in the JSON description to be applied in JS. If there was a way to make sure this "`eval`" is safe, it could mean faster data updates. (I don't know how unsafe `eval` is in the browser, specifically).

If accessors are applied only or mostly inside JS, that could reduce code duplication across languages. Though I also understand if you think parsing a function from the JSON attributes is too hacky.

**Picking**

If an entire table is sent to JS from the binding, picking is able to return metadata along with the feature, rather than solely its geometry.

**Less custom binding code**

Arrow is pretty widespread across languages, so i.e. in Python a user working with Pandas could pass an Arrow table to Deck.gl frictionlessly, and the binding wouldn't need custom code to serialize into buffers.

**GPUTable module?**

If the GPUTable module PR becomes merged and Arrow is used in the GPU, it's not that much of a stretch to use Arrow in JS internally for some things.

**Streaming**

Built-in support in Arrow for chunked table streaming.
